lib/OpenQA/WebAPI/Controller/Step.pm

```diff
@@ -91,6 +94,34 @@ sub view ($self) {
     $self->viewimg;
 }
 
+sub _determine_needles_dir_for_job ($self, $job) {
+    return undef unless $self->app->config->{'scm git'}->{checkout_needles_sha} eq 'yes';
+    return undef unless my $needle_dirs = realpath($job->needle_dir);
+    my $settings = $job->settings;
+    return ($needle_dirs, $settings->single({key => 'NEEDLES_DIR'}) // $settings->single({key => 'CASEDIR'}));
+}
+
+sub _create_tmpdir_for_needles_refspec ($self, $job) {
+    my ($needle_dirs, $needles_dir_var) = $self->_determine_needles_dir_for_job($job);
+    return undef unless $needles_dir_var;
+    my $needles_url = Mojo::URL->new($needles_dir_var->value);
+    return undef unless $needles_url->scheme;
+    my $needles_ref = $needles_url->fragment;
+    eval {
+        my $vars = decode_json(path($job->result_dir, 'vars.json')->slurp);
+        $needles_ref = $vars->{NEEDLES_GIT_HASH} if ref $vars eq 'HASH';
+    };
+    chomp $needles_ref;
+    return undef unless $needles_ref;
+    return undef unless run_cmd_with_log ['git', '-C', $needle_dirs, 'fetch', '--depth', 1, 'origin', $needles_ref];
```

Comment:
@sdclarke We have not forgotten about this.

We are currently working on some git features and fixes, and I just wanted to note that we have a `git_auto_clone` feature now that will fetch any requested ref on the webui (branch or sha) in a minion job before the job starts running. (Note that it only works if it is the same origin and not some fork.)

That would make this `fetch` here not necessary anymore and also avoid the need for a lockfile.
The sha should always be in `NEEDLES_GIT_HASH` so the `rev-parse` also seems unnecessary.

Would you consider to use the `git_auto_clone` feature in order to not have to implement that part here?

I'm not sure if there are any cases which `git_auto_clone`  wouldn't cover. @Martchus said there might be.

Reply:
Note that `git_auto_clone` only helps if a Git-URL is specified as `NEEDLES_DIR` (or `CASEDIR` which is relevant if needles are part of the tests).

Even then it is not guaranteed to fetch the refs the test actually runs on. The test might specify a branch name (or no branch which means "default branch") and when the test is finally executed on the worker host the ref that branch points to might differ from the ref checked out by the web UI at the time the test was scheduled. 

Additionally, we probably want to have some cleanup for the fetching done by the `git_auto_clone` feature. So the ref might have also already been cleaned up.

So I think we need that fetch in any case.

---

> That would make this fetch here not necessary anymore and also avoid the need for a lockfile.

I don't think we need a lockfile here. We can probably arrange it so that it would not matter if any other commands are executed in the middle. (The use of `FETCH_HEAD` definitely needs to go way but it can probably be replaced with whatever we specify on the `fetch` command.)

> The sha should always be in NEEDLES_GIT_HASH so the rev-parse also seems unnecessary.

That's true. So we could (and probably should) get rid of it. If one specifies a branch name via `NEEDLES_DIR` the `NEEDLES_GIT_HASH` still has the concrete sha.

Reply:
> Note that `git_auto_clone` only helps if a Git-URL is specified as `NEEDLES_DIR` (or `CASEDIR` which is relevant if needles are part of the tests).

* The PR description says that it is about CASEDIR/NEEDLES_DIR being set.
* And if it is not set, there is no certain ref to fetch.
* In the near future, `git_auto_clone` will also automatically update clones if CASEDIR/NEEDLES_DIR is empty. It will just update the checkout branch in the checked out directory.

And this is all we need, no?
What else needs to be fetched that git_auto_clone does not fetch?

The only thing could be forks, but as far as I can see this PR also doesn't handle forks.


> Even then it is not guaranteed to fetch the refs the test actually runs on. The test might specify a branch name (or no branch which means "default branch") and when the test is finally executed on the worker host the ref that branch points to might differ from the ref checked out by the web UI at the time the test was scheduled.

What's checked out on the webui does not matter. The important thing is that it has been fetched at some point, so we can do a `git show` on it, right?

> Additionally, we probably want to have some cleanup for the fetching done by the `git_auto_clone` feature. So the ref might have also already been cleaned up.

Yes. I'm not sure what git already offers there and how to know what we can cleanup.

Reply:
> What's checked out on the webui does not matter. The important thing is that it has been fetched at some point, so we can do a git show on it, right?

I meant "fetched by the web UI". And I meant the following sequence of events:

1. A test is scheduled pointing to a branch, e.g. just the default branch.
2. The web UI fetches the latest ref of that branch because `git_auto_clone = 1` is configured.
3. Someone pushes additional commits to that branch or even force-pushes there.
4. The test is finally assigned to a worker and that worker now fetches a different ref than the web UI in step 2 so `NEEDLES_GIT_HASH` points to a ref that is not yet checked out (despite `git_auto_clone = 1`).
5. Someone wants to display the candidate needles for that job. We need to fetch the missing ref.

And this is not very contrived. That someone specifies a branch where many commits happen (e.g. the default branch of some repo with many contributors) is probably a common use case. That it can take between the scheduling of a job and the time it actually gets executed is also something we need to expect.